fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java

```diff
@@ -46,15 +45,22 @@ public LocalDate getLastRepaymentDate(final LoanApplicationTerms loanApplication
 
         final int numberOfRepayments = loanApplicationTerms.getNumberOfRepayments();
 
-        RepaymentStartDateType repaymentStartDateType = loanApplicationTerms.getRepaymentStartDateType();
+        LocalDate lastRepaymentDate = loanApplicationTerms.getRepaymentStartDate();
+        final LocalDate maxDateForFixedLength = loanApplicationTerms.calculateMaxDateForFixedLength();
 
-        LocalDate lastRepaymentDate = RepaymentStartDateType.DISBURSEMENT_DATE.equals(repaymentStartDateType)
-                ? loanApplicationTerms.getExpectedDisbursementDate()
-                : loanApplicationTerms.getSubmittedOnDate();
         boolean isFirstRepayment = true;
         for (int repaymentPeriod = 1; repaymentPeriod <= numberOfRepayments; repaymentPeriod++) {
             lastRepaymentDate = generateNextRepaymentDate(lastRepaymentDate, loanApplicationTerms, isFirstRepayment);
             isFirstRepayment = false;
+            // Fixed Length validation - When the Number of Repayments is greater than Fixed Length
+            if (maxDateForFixedLength != null && DateUtils.isAfter(lastRepaymentDate, maxDateForFixedLength)) {
+                lastRepaymentDate = maxDateForFixedLength;
+                break;
```

Comment:
Would you mind moving this condition out of the cycle, it does not require do to anything there. After the cycle finished you will have the `lastRepaymentDate` set, which is the due date of the last installment and after you can decide whether it needs to be changed or not:
if `lastRepaymentDate` is not equal with the `fixedLengthDate`, override it with the `fixedLengthDate`.

Reply:
Done!